### DISCUSSION TITLE
problems compiling graphics with llvm-mos-sdk

### DISCUSSION THREAD

**Discussion by vruumllc** [picocomputer/community#83] [Q&A] [answered]
Hi, I tried installing the llvm-mos compiler using the link in the llvm-mos-sdk, and was able to build and run simple c and c++ examples that didn't use the rp6502.h file.

When trying to initialize graphics for the RP6502, however, I had linker failure with the xregn() function missing.

In the rp6502.h file of the sdk, the xreg() function has been renamed to xregn(), and the xreg_vga_mode macro is missing.

The xreg() function has also renamed to xregn() in the rp6502.h file from the downloaded compiler, and the xreg_vga_mode macro is present, but it is defined to use xreg()!

I tied a bunch of stuff but never could get my program to compile.  Any suggestions?

here are snippets from the conflicting header files:
--------------------------------------------------------------------------------------------------------------------------------------------------------------
`In rp6502.h from the SDK, it looks like this:
....
/* C API for the operating system. */

int xregn(char device, char channel, unsigned char address, unsigned count,
          ...);
int phi2(void);
int codepage(void);
long lrand(void);
int stdin_opt(unsigned long ctrl_bits, unsigned char str_length);
int read_xstack(void *buf, unsigned count, int fildes);
int read_xram(unsigned buf, unsigned count, int fildes);
int write_xstack(const void *buf, unsigned count, int fildes);
int write_xram(unsigned buf, unsigned count, int fildes);

/* Temporary home for standard library headers. */
....
-------------------------------------------------------------------------------------------------------------------------------------------------------------------
In the rp6502.h in the downloaded compiler, it looks like this:

....
//* C API for the operating system. */

int xregn(char device, char channel, unsigned char address, unsigned char count,
          ...);
int phi2(void);
int codepage(void);
long lrand(void);
int stdin_opt(unsigned long ctrl_bits, unsigned char str_length);

/* XREG location helpers */

#define xreg_ria_keyboard(...) xreg(0, 0, 0, __VA_ARGS__)
#define xreg_ria_mouse(...) xreg(0, 0, 1, __VA_ARGS__)
#define xreg_vga_canvas(...) xreg(1, 0, 0, __VA_ARGS__)
#define xreg_vga_mode(...) xreg(1, 0, 1, __VA_ARGS__)

/* XRAM structure helpers */
....
-------------------------------------------------------------------------------------------------------------------------------------------------------------------

**Comment by rumbledethumps** [comment]
xregn is not a renamed xreg. it has a different prototype. have you built llvm-mos-sdk or are you using a binary distribution and did you try the other?

**Reply by vruumllc** [reply]
Interesting.  
Yeah, I tried compiling the SDK, and tried downloading the binary.  Like I said, they seem different. I couldn't get either to work.  I'll try again.

**Reply by vruumllc** [reply]
Ok, I deleted everything and then did this:

(created my target directory for the SDK build)
'mkdir /home/tonyvr/llvm-mos'

(followed the directions on  https://github.com/picocomputer/llvm-mos-sdk?tab=readme-ov-file#build-and-install-llvm-mos-sdk )
'$ git clone https://github.com/llvm-mos/llvm-mos-sdk.git'
'$ cd llvm-mos-sdk'
'$ mkdir build'
'$ cd build'
'$ cmake -G "Ninja" -DCMAKE_INSTALL_PREFIX=/home/tonyvr/llvm-mos ..'
'$ ninja install'

(this all went fine, and I have a fresh build of the SDK in /home/tonyvr/llvm-mos)

(I then started VSCODE in my te_cpp project directory, created from https://github.com/picocomputer/vscode-llvm-mos template)
(I had previously added .vscode/settings.json file, with path to the llvm-mos compiler at /home/tonyvr/llvm-mos/bin)
(A generic hello_world.cpp using printf() was previously compiled with this setup and ran fine)

(Compiling my new project, te_cpp, failed with message:)

[build] ld.lld: error: undefined symbol: xregn(char, char, unsigned char, unsigned int, ...)

**Reply by rumbledethumps** [reply]
You can try the code I built for the video.
https://github.com/rumbledethumps/examples-llvm-mos

**Reply by vruumllc** [reply]
OK, Your stuff compiles, but the Mandelbrot hangs (incompatible with my firmware???). Palette works fine.

I changed my test to be a '.c' file and it now compiles, and seems to be running, but it doesn't do anything except init bitmap mode, so hard to tell yet..

So... it seems I get the linker error with CPP but not C.

**Reply by vruumllc** [reply]
Ah! I was using download binary release of llvm-mos during Mandelbrot test.  I tried switching to the one I built with the SDK and that works fine!!! 

I just tried running my CPP version again with SDK version, and it still gets xregn() linker error. So CPP still a problem.

**Comment by rumbledethumps** [comment]
https://github.com/llvm-mos/llvm-mos-sdk/pull/305

**Reply by vruumllc** [reply]
Cool, sounds like you figured out the problem and have a solution. Fast work. Thanks!

**Reply by vruumllc** [reply]
I saw you checked in the fix (I hope you weren't up all night).

CPP graphics now compile and run fine if you rebuild the compiler with SDK, per directions above.
I'm currently porting my mode1 (text mode) display library to C++ and will upload a demo shortly.

Thanks!

**Reply by vruumllc** [reply, marked as answer]
I uploaded new CPP demo to:
https://github.com/orgs/picocomputer/discussions/84

**Reply by rumbledethumps** [reply]
Note that SDK v12.0.0 was released today. That'll have the C++ patch.
